File: conformance/tests/merge-locals-prop-updates/MergeLocalsPropUpdates.runar.rb
require 'runar'

# MergeLocalsPropUpdates -- the branch-merge shape crossed with property
# mutation: one method that BOTH merges two locals across an `if` AND writes
# contract properties from the merged results. No contract in the repo did
# both before this one.
class MergeLocalsPropUpdates < Runar::StatefulSmartContract
  prop :a, Bigint
  prop :b, Bigint
  prop :total, Bigint

  def initialize(a, b, total)
    super(a, b, total)
    @a = a
    @b = b
    @total = total
  end

  runar_public amount: Bigint, to_first: Bigint
  def settle(amount, to_first)
    na = @a
    nb = @b
    if to_first > 0
      na = na + amount
    else
      nb = nb + amount
    end
    @a = na
    @b = nb
    @total = na + nb
    add_output(1000, @a, @b, @total)
  end
end

File: conformance/tests/merge-locals-prop-updates/MergeLocalsPropUpdates.runar.py
from runar import (
    StatefulSmartContract, Bigint, public,
)


class MergeLocalsPropUpdates(StatefulSmartContract):
    """The branch-merge shape crossed with property mutation: one method that
    BOTH merges two locals across an `if` AND writes contract properties from
    the merged results. No contract in the repo did both before this one."""

    a: Bigint
    b: Bigint
    total: Bigint

    def __init__(self, a: Bigint, b: Bigint, total: Bigint):
        super().__init__(a, b, total)
        self.a = a
        self.b = b
        self.total = total

    @public
    def settle(self, amount: Bigint, to_first: Bigint):
        na = self.a
        nb = self.b
        if to_first > 0:
            na = na + amount
        else:
            nb = nb + amount
        self.a = na
        self.b = nb
        self.total = na + nb
        self.add_output(1000, self.a, self.b, self.total)
